vortex-scalar/src/primitive.rs

```diff
@@ -270,3 +272,70 @@ impl From<usize> for Scalar {
         Scalar::primitive(value as u64, Nullability::NonNullable)
     }
 }
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum BinaryNumericOperator {
+    Add,
+    Sub,
+    Mul,
+    Div,
+    // Missing from arrow-rs:
+    // Min,
+    // Max,
+    // Pow,
+}
+
+impl PrimitiveScalar<'_> {
+    /// Apply the (checked) operator to self and other using SQL-style null semantics.
+    ///
+    /// If the operation overflows, Ok(None) is returned.
+    ///
+    /// If the types are incompatible (ignoring nullability), an error is returned.
+    ///
+    /// If either value is null, the result is null.
+    pub fn checked_numeric_operator(
+        self,
+        other: PrimitiveScalar<'_>,
+        op: BinaryNumericOperator,
+    ) -> VortexResult<Option<Scalar>> {
+        if !self.dtype().eq_ignore_nullability(other.dtype()) {
+            vortex_bail!("types must match: {} {}", self.dtype(), other.dtype());
+        }
+
+        let nullability = self.dtype().nullability();
+
+        Ok(match_each_native_ptype!(
+            self.ptype(),
+            integral: |$P| {
+                let lhs = self.typed_value::<$P>();
+                let rhs = other.typed_value::<$P>();
+                match (lhs, rhs) {
+                    (_, None) | (None, _) => Some(Scalar::null(self.dtype().clone().as_nullable())),
```

Comment:
This is the bug I think is still here. If `(_, None)` is true, and lhs is non-nullable, then you're going to try to create a Scalar::null that's non-nullable

Reply:
This case is correct (b/c of the as_nullable) but now every case uses the same (least viable) nullability.